src/dmd/doc.d

```diff
@@ -1823,6 +1833,32 @@ extern (C++) bool isDitto(const(char)* comment)
     return false;
 }
 
+/*****************************************
+ * Return true if the given character is white space.
+ */
+private bool isWhitespace(const(char) c)
+{
+// TODO: unicode whitespace: Zs.
+    return c == ' ' || c == '\t' || c == '\r' || c == '\n';
+}
+
+/*****************************************
+ * Return true if the given character is punctuation.
+ */
+private bool isPunctuation(const(char) c)
+{
+// TODO: unicode punctuation: Pc, Pd, Pe, Pf, Pi, Po, or Ps.
+// However, unicode punctuation should not be included for Markdown backslash
+// escapes, so when it's implemented be sure to make the unicode punctuation
+// check optional via a parameter
+    return c == '!' || c == '"' || c == '#' || c == '$' || c == '%' || c == '&'
```

Comment:
Use a static array of booleans, indexed by char (as ubyte), for O(1) lookups:

```d
static immutable bool[char.max] punctLUT= ['!' : true, ',': true, /*etc*/ ];
```

Only `true` members needs to be initialized. Refactoring for unicode, if necessary will be easy. (if ... > 0x80)...
Also these are rather called symbols but let's  not bikeshed for now.

Reply:
I think that there's been a slight misunderstanding: you finally use an associative array but i proposed a a simple static immutable array, meaning that the data don't need to be initialized.
Look at the difference here: https://godbolt.org/g/VMrxDj